prettyprinter.go

```diff
@@ -0,0 +1,158 @@
+package quamina
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+)
+
+// printer is an interface used to generate representations of Quamina data structures to facilitate
+// debugging and optimization. It's an interface rather than a type so that a null implementation can
+// be provided for production that should incur very little performance cost.
+type printer interface {
```

Comment:
I'd suggest following what Google did when they developed the `slog` logging package. Instead of going with an `interface` type, they used a concrete struct with a specific `interface` handler type to avoid allocations:  https://youtu.be/8rnI2xLrdeM?feature=shared&t=1336 (starts at that conversation)

Reply:
tl;dr

```go
// inspired by https://pkg.go.dev/golang.org/x/exp/slog#Logger
type Handler interface {
  Enabled() bool
  Handle(...)
}

type Printer struct {
  handler handler
}

func New() *Printer {...}
```

Reply:
Also, I agree about documenting this stuff. Probably best to create a `DEVELOPING.md` or some such; is there a standard name for this kind of thing?

Reply:
If this implementation is purely for developing, `CONTRIBUTING.md` might be common

Reply:
Left a comment in the place where I think alloc might happen? Ideally, we'd have a benchmark which shows the before/after here?

Reply:
Can't see alloc changes in this though: https://github.com/timbray/quamina/actions/runs/9324906941?pr=311

Reply:
Yes, CONTRIBUTING is the place.

Reply:
I added a description to CONTRIBUTING in the latest commit. If that's OK I think we can resolve this conversation?